exonum-java-binding/testing/src/main/java/com/exonum/binding/test/runtime/ServiceArtifactBuilder.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2019 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exonum.binding.test.runtime;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.util.Arrays.asList;
+
+import com.google.common.io.ByteStreams;
+import java.io.BufferedOutputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintWriter;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.jar.Attributes.Name;
+import java.util.jar.JarOutputStream;
+import java.util.jar.Manifest;
+import java.util.stream.Stream;
+import java.util.zip.ZipEntry;
+
+/**
+ * A builder of service artifacts in PF4J format. Intended to be used in various integration tests.
+ * Allows to create malformed artifacts (with missing classes, extensions, incorrect metadata).
+ */
+public final class ServiceArtifactBuilder {
+
+  static final String PLUGIN_ID_ATTRIBUTE_NAME = "Plugin-Id";
+  static final String EXTENSIONS_INDEX_NAME = "META-INF/extensions.idx";
+
+  private final Manifest manifest;
+  private final Set<Class<?>> artifactClasses;
+  /** Entries to put into the list of extensions. */
+  private final LinkedHashSet<String> extensions;
+
+  /** Creates a new artifact builder with empty manifest, no classes and no extensions. */
+  public ServiceArtifactBuilder() {
+    manifest = new Manifest();
+    artifactClasses = new HashSet<>();
+    extensions = new LinkedHashSet<>();
+
+    manifest.getMainAttributes().put(Name.MANIFEST_VERSION, "1.0");
+  }
+
+  /**
+   * Sets a PF4J plugin identifier.
+   * @param pluginId a plugin id to set; in Exonum it must have a certain format, but this class
+   *     allows any to be set
+   * @see <a href="https://pf4j.org/doc/plugins.html#how-plugin-metadata-is-defined">Plugin meta data</a>
+   */
+  public ServiceArtifactBuilder setPluginId(String pluginId) {
+    return setManifestEntry(PLUGIN_ID_ATTRIBUTE_NAME, pluginId);
+  }
+
+  /**
+   * Sets a manifest entry in the {@linkplain Manifest main section} of the manifest.
+   * @param name an attribute name
+   * @param value an attribute value
+   * @throws IllegalArgumentException if the attribute name is invalid
```

Comment:
Not clear what is the criteria for `valid` name. Specify the format above?

Reply:
That's manifest format restrictions. I can link them, I hope.